File: app/src/main/kotlin/tech/mmarca/openvitals/data/local/beverage/BeverageEntity.kt
package tech.mmarca.openvitals.data.local.beverage

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import tech.mmarca.openvitals.domain.insights.BeverageNutritionDefaults
import tech.mmarca.openvitals.domain.insights.CaffeineHealthDrinkCatalog
import tech.mmarca.openvitals.domain.model.CaffeineCatalogItem
import tech.mmarca.openvitals.domain.model.CaffeineSourceCategory
import tech.mmarca.openvitals.domain.model.CustomHydrationDrink
import tech.mmarca.openvitals.domain.model.NutritionNutrient

@Entity(tableName = "beverages")
data class BeverageEntity(
    @PrimaryKey val id: String,
    val name: String,
    val category: String?,
    @ColumnInfo(name = "volume_milliliters") val volumeMilliliters: Double,
    @ColumnInfo(name = "hydration_multiplier") val hydrationMultiplier: Double,
    @ColumnInfo(name = "is_preloaded") val isPreloaded: Boolean,
    @ColumnInfo(name = "is_deleted") val isDeleted: Boolean = false,
    @ColumnInfo(name = "sort_order") val sortOrder: Int,
    @ColumnInfo(name = "energy_kcal") val energyKcal: Double? = null,
    @ColumnInfo(name = "protein_grams") val proteinGrams: Double? = null,
    @ColumnInfo(name = "total_carbohydrate_grams") val totalCarbohydrateGrams: Double? = null,
    @ColumnInfo(name = "total_fat_grams") val totalFatGrams: Double? = null,
    @ColumnInfo(name = "dietary_fiber_grams") val dietaryFiberGrams: Double? = null,
    @ColumnInfo(name = "sugar_grams") val sugarGrams: Double? = null,
    @ColumnInfo(name = "saturated_fat_grams") val saturatedFatGrams: Double? = null,
    @ColumnInfo(name = "sodium_grams") val sodiumGrams: Double? = null,
    @ColumnInfo(name = "potassium_grams") val potassiumGrams: Double? = null,
    @ColumnInfo(name = "calcium_grams") val calciumGrams: Double? = null,
    @ColumnInfo(name = "caffeine_grams") val caffeineGrams: Double? = null,
) {
    fun toDomain(): CustomHydrationDrink =
        CustomHydrationDrink(
            id = id,
            name = name,
            volumeMilliliters = volumeMilliliters,
            hydrationMultiplier = hydrationMultiplier,
            nutrientValues = nutrientValues(),
            category = category?.let { runCatching { CaffeineSourceCategory.valueOf(it) }.getOrNull() },
            isPreloaded = isPreloaded,
        )

    private fun nutrientValues(): Map<NutritionNutrient, Double> = buildMap {
        putPositive(NutritionNutrient.ENERGY, energyKcal)
        putPositive(NutritionNutrient.PROTEIN, proteinGrams)
        putPositive(NutritionNutrient.TOTAL_CARBOHYDRATE, totalCarbohydrateGrams)
        putPositive(NutritionNutrient.TOTAL_FAT, totalFatGrams)
        putPositive(NutritionNutrient.DIETARY_FIBER, dietaryFiberGrams)
        putPositive(NutritionNutrient.SUGAR, sugarGrams)
        putPositive(NutritionNutrient.SATURATED_FAT, saturatedFatGrams)
        putPositive(NutritionNutrient.SODIUM, sodiumGrams)
        putPositive(NutritionNutrient.POTASSIUM, potassiumGrams)
        putPositive(NutritionNutrient.CALCIUM, calciumGrams)
        putPositive(NutritionNutrient.CAFFEINE, caffeineGrams)
    }

    private fun MutableMap<NutritionNutrient, Double>.putPositive(
        nutrient: NutritionNutrient,
        value: Double?,
    ) {
        if (value != null && value > 0.0 && value.isFinite()) {
            put(nutrient, value)
        }
    }

    companion object {
        fun preloadedDefaults(): List<BeverageEntity> =
            listOf(waterDefault()) + CaffeineHealthDrinkCatalog.items
                .asSequence()
                .filter { item -> item.defaultServingMilliliters != null }
                .filterNot { item -> item.category == CaffeineSourceCategory.SUPPLEMENT }
                .mapIndexed { index, item -> item.toPreloadedEntity(index + 1) }
                .toList()

        private fun waterDefault(): BeverageEntity =
            fromDomain(
                drink = CustomHydrationDrink(
                    id = "openvitals-water",
                    name = "Water",
                    volumeMilliliters = 100.0,
                    hydrationMultiplier = 1.0,
                    nutrientValues = emptyMap(),
                    category = null,
                    isPreloaded = true,
                ),
                sortOrder = 0,
                isPreloaded = true,
                category = null,
            )

        fun fromDomain(
            drink: CustomHydrationDrink,
            sortOrder: Int,
            isPreloaded: Boolean = drink.isPreloaded,
            category: CaffeineSourceCategory? = drink.category,
        ): BeverageEntity =
            BeverageEntity(
                id = drink.id,
                name = drink.name,
                category = category?.name,
                volumeMilliliters = drink.volumeMilliliters,
                hydrationMultiplier = drink.hydrationMultiplier,
                isPreloaded = isPreloaded,
                sortOrder = sortOrder,
                energyKcal = drink.nutrientValues[NutritionNutrient.ENERGY],
                proteinGrams = drink.nutrientValues[NutritionNutrient.PROTEIN],
                totalCarbohydrateGrams = drink.nutrientValues[NutritionNutrient.TOTAL_CARBOHYDRATE],
                totalFatGrams = drink.nutrientValues[NutritionNutrient.TOTAL_FAT],
                dietaryFiberGrams = drink.nutrientValues[NutritionNutrient.DIETARY_FIBER],
                sugarGrams = drink.nutrientValues[NutritionNutrient.SUGAR],
                saturatedFatGrams = drink.nutrientValues[NutritionNutrient.SATURATED_FAT],
                sodiumGrams = drink.nutrientValues[NutritionNutrient.SODIUM],
                potassiumGrams = drink.nutrientValues[NutritionNutrient.POTASSIUM],
                calciumGrams = drink.nutrientValues[NutritionNutrient.CALCIUM],
                caffeineGrams = drink.nutrientValues[NutritionNutrient.CAFFEINE],
            )

        private fun CaffeineCatalogItem.toPreloadedEntity(sortOrder: Int): BeverageEntity =
            fromDomain(
                drink = CustomHydrationDrink(
                    id = "$BeveragePresetIdPrefix$id",
                    name = name,
                    volumeMilliliters = defaultServingMilliliters ?: 240.0,
                    hydrationMultiplier = 1.0,
                    nutrientValues = BeverageNutritionDefaults.nutrientValuesFor(this),
                    category = category,
                    isPreloaded = true,
                ),
                sortOrder = sortOrder,
                isPreloaded = true,
                category = category,
            )

        private const val BeveragePresetIdPrefix = "caffeinehealth-"
    }
}

File: lib/data/local/beverage/beverage_entity.dart
import '../../../domain/insights/caffeine_health_drink_catalog.dart';
import '../../../domain/model/caffeine_models.dart';
import '../../../domain/model/nutrition_models.dart';

/// Row model for the `beverages` table.
///
/// This mirrors the Kotlin Room `BeverageEntity` one-to-one and doubles as the
/// drift row class (via `@UseRowClass`), so its constructor parameters must
/// match the table columns exactly.
class BeverageEntity {
  const BeverageEntity({
    required this.id,
    required this.name,
    required this.category,
    required this.volumeMilliliters,
    required this.hydrationMultiplier,
    required this.isPreloaded,
    this.isDeleted = false,
    required this.sortOrder,
    this.energyKcal,
    this.proteinGrams,
    this.totalCarbohydrateGrams,
    this.totalFatGrams,
    this.dietaryFiberGrams,
    this.sugarGrams,
    this.saturatedFatGrams,
    this.sodiumGrams,
    this.potassiumGrams,
    this.calciumGrams,
    this.caffeineGrams,
  });

  final String id;
  final String name;
  final String? category;
  final double volumeMilliliters;
  final double hydrationMultiplier;
  final bool isPreloaded;
  final bool isDeleted;
  final int sortOrder;
  final double? energyKcal;
  final double? proteinGrams;
  final double? totalCarbohydrateGrams;
  final double? totalFatGrams;
  final double? dietaryFiberGrams;
  final double? sugarGrams;
  final double? saturatedFatGrams;
  final double? sodiumGrams;
  final double? potassiumGrams;
  final double? calciumGrams;
  final double? caffeineGrams;

  CustomHydrationDrink toDomain() => CustomHydrationDrink(
        id: id,
        name: name,
        volumeMilliliters: volumeMilliliters,
        hydrationMultiplier: hydrationMultiplier,
        nutrientValues: _nutrientValues(),
        category: category == null
            ? null
            : CaffeineSourceCategory.fromStorage(category!),
        isPreloaded: isPreloaded,
      );

  Map<NutritionNutrient, double> _nutrientValues() {
    final values = <NutritionNutrient, double>{};
    void putPositive(NutritionNutrient nutrient, double? value) {
      if (value != null && value > 0.0 && value.isFinite) {
        values[nutrient] = value;
      }
    }

    putPositive(NutritionNutrient.energy, energyKcal);
    putPositive(NutritionNutrient.protein, proteinGrams);
    putPositive(NutritionNutrient.totalCarbohydrate, totalCarbohydrateGrams);
    putPositive(NutritionNutrient.totalFat, totalFatGrams);
    putPositive(NutritionNutrient.dietaryFiber, dietaryFiberGrams);
    putPositive(NutritionNutrient.sugar, sugarGrams);
    putPositive(NutritionNutrient.saturatedFat, saturatedFatGrams);
    putPositive(NutritionNutrient.sodium, sodiumGrams);
    putPositive(NutritionNutrient.potassium, potassiumGrams);
    putPositive(NutritionNutrient.calcium, calciumGrams);
    putPositive(NutritionNutrient.caffeine, caffeineGrams);
    return values;
  }

  BeverageEntity copyWith({
    String? id,
    String? name,
    String? category,
    double? volumeMilliliters,
    double? hydrationMultiplier,
    bool? isPreloaded,
    bool? isDeleted,
    int? sortOrder,
  }) =>
      BeverageEntity(
        id: id ?? this.id,
        name: name ?? this.name,
        category: category ?? this.category,
        volumeMilliliters: volumeMilliliters ?? this.volumeMilliliters,
        hydrationMultiplier: hydrationMultiplier ?? this.hydrationMultiplier,
        isPreloaded: isPreloaded ?? this.isPreloaded,
        isDeleted: isDeleted ?? this.isDeleted,
        sortOrder: sortOrder ?? this.sortOrder,
        energyKcal: energyKcal,
        proteinGrams: proteinGrams,
        totalCarbohydrateGrams: totalCarbohydrateGrams,
        totalFatGrams: totalFatGrams,
        dietaryFiberGrams: dietaryFiberGrams,
        sugarGrams: sugarGrams,
        saturatedFatGrams: saturatedFatGrams,
        sodiumGrams: sodiumGrams,
        potassiumGrams: potassiumGrams,
        calciumGrams: calciumGrams,
        caffeineGrams: caffeineGrams,
      );

  static const String beveragePresetIdPrefix = 'caffeinehealth-';

  static List<BeverageEntity> preloadedDefaults() {
    final presets = CaffeineHealthDrinkCatalog.beveragePresets();
    return <BeverageEntity>[
      ..._waterDefaults(),
      for (var index = 0; index < presets.length; index++)
        BeverageEntity.fromDomain(
          drink: presets[index],
          sortOrder: index + 2,
          isPreloaded: true,
          category: presets[index].category,
        ),
    ];
  }

  static List<BeverageEntity> _waterDefaults() => <BeverageEntity>[
        _waterDefault(
          id: 'openvitals-still-water',
          name: 'Still water',
          sortOrder: 0,
        ),
        _waterDefault(
          id: 'openvitals-gasified-water',
          name: 'Gasified water',
          sortOrder: 1,
        ),
      ];

  static BeverageEntity _waterDefault({
    required String id,
    required String name,
    required int sortOrder,
  }) =>
      BeverageEntity.fromDomain(
        drink: CustomHydrationDrink(
          id: id,
          name: name,
          volumeMilliliters: 100.0,
          hydrationMultiplier: 1.0,
          nutrientValues: const <NutritionNutrient, double>{},
          category: CaffeineSourceCategory.water,
          isPreloaded: true,
        ),
        sortOrder: sortOrder,
        isPreloaded: true,
        category: CaffeineSourceCategory.water,
      );

  /// Mirrors the Kotlin `fromDomain`, whose `isPreloaded`/`category`
  /// parameters default to `drink.isPreloaded`/`drink.category`. Every call
  /// site passes these explicitly, so the `?? drink.*` fallback only supplies
  /// the Kotlin default when a caller omits them.
  static BeverageEntity fromDomain({
    required CustomHydrationDrink drink,
    required int sortOrder,
    bool? isPreloaded,
    CaffeineSourceCategory? category,
  }) {
    final resolvedCategory = category ?? drink.category;
    final nutrients = drink.nutrientValues;
    return BeverageEntity(
      id: drink.id,
      name: drink.name,
      category: resolvedCategory?.storageName,
      volumeMilliliters: drink.volumeMilliliters,
      hydrationMultiplier: drink.hydrationMultiplier,
      isPreloaded: isPreloaded ?? drink.isPreloaded,
      sortOrder: sortOrder,
      energyKcal: nutrients[NutritionNutrient.energy],
      proteinGrams: nutrients[NutritionNutrient.protein],
      totalCarbohydrateGrams: nutrients[NutritionNutrient.totalCarbohydrate],
      totalFatGrams: nutrients[NutritionNutrient.totalFat],
      dietaryFiberGrams: nutrients[NutritionNutrient.dietaryFiber],
      sugarGrams: nutrients[NutritionNutrient.sugar],
      saturatedFatGrams: nutrients[NutritionNutrient.saturatedFat],
      sodiumGrams: nutrients[NutritionNutrient.sodium],
      potassiumGrams: nutrients[NutritionNutrient.potassium],
      calciumGrams: nutrients[NutritionNutrient.calcium],
      caffeineGrams: nutrients[NutritionNutrient.caffeine],
    );
  }
}
